Authentication 

In [1]:
from google.colab import auth
auth.authenticate_user()
print('Authenticated')

Authenticated


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import plotly.express as px
import plotly as py
import plotly.graph_objs as go
from plotly.offline import init_notebook_mode, iplot

Connecting to Google Bigquery public data “iowa_liquor_sales”, extracting data since 2020 and saving it to pandas.DataFrame

In [5]:
%%bigquery --project myproject-293510 test_task_data
SELECT * FROM `bigquery-public-data.iowa_liquor_sales.sales`  where extract(year from date)=2020 order by date desc

Displaying head of dataframe

In [8]:
test_task_data.head()

invoice_and_item_number        date  ... volume_sold_liters volume_sold_gallons
0         INV-30655500013  2020-09-30  ...               0.05                0.01
1         INV-30661000005  2020-09-30  ...               0.75                0.19
2         INV-30659500001  2020-09-30  ...               9.00                2.37
3         INV-30652600028  2020-09-30  ...               0.75                0.19
4         INV-30650100014  2020-09-30  ...               0.10                0.02

[5 rows x 24 columns]

In [6]:
test_task_data.columns

Index(['invoice_and_item_number', 'date', 'store_number', 'store_name',
       'address', 'city', 'zip_code', 'store_location', 'county_number',
       'county', 'category', 'category_name', 'vendor_number', 'vendor_name',
       'item_number', 'item_description', 'pack', 'bottle_volume_ml',
       'state_bottle_cost', 'state_bottle_retail', 'bottles_sold',
       'sale_dollars', 'volume_sold_liters', 'volume_sold_gallons'],
      dtype='object')

In [9]:
test_task_data.dtypes

invoice_and_item_number     object
date                        object
store_number                object
store_name                  object
address                     object
city                        object
zip_code                    object
store_location              object
county_number               object
county                      object
category                    object
category_name               object
vendor_number               object
vendor_name                 object
item_number                 object
item_description            object
pack                         int64
bottle_volume_ml             int64
state_bottle_cost          float64
state_bottle_retail        float64
bottles_sold                 int64
sale_dollars               float64
volume_sold_liters         float64
volume_sold_gallons        float64
dtype: object

Number and list of unique cities in data

In [10]:
print(test_task_data['city'].nunique())
test_task_data['city'].unique()

431


array(['Ankeny', 'Glenwood', 'Ottumwa', 'Sigourney', 'Fairfield',
       'Council Bluffs', 'Spirit Lake', 'Ames', 'Washington', 'Keokuk',
       'Des Moines', 'Cedar Rapids', 'Keosauqua', 'Spencer', 'Davenport',
       'Paullina', 'Forest City', 'Pella', 'Algona', 'Britt',
       'Emmetsburg', 'Hartley', 'West Des Moines', 'Iowa City',
       'Oskaloosa', 'Waterloo', 'Milford', 'Fort Madison', 'Arnolds Park',
       'Estherville', 'Windsor Heights', 'Marcus', 'Belmond', 'Okoboji',
       'Winfield', 'Pacific Junction', 'Correctionville', 'Bloomfield',
       'Bancroft', 'Eddyville', 'West Point', 'Lake Mills', 'Cedar Falls',
       'Sanborn', 'Marshalltown', 'Farmington', 'Kingsley', 'Hiawatha',
       'Wilton', 'West Liberty', 'Colfax', 'Alta', 'Garner', 'Clarion',
       'Swea City', 'West Bend', 'Durant', 'Primghar', 'Griswold',
       'Wesley', 'Bettendorf', 'Corwith', 'Coralville', 'Urbandale',
       'Eagle Grove', 'Lake Park', 'Ruthven', 'Osceola', 'Buffalo Center',
       'Rems

There are 431 cities in explored data


In [11]:
test_task_data['city'].value_counts()

Des Moines         154389
Cedar Rapids       125613
Davenport           82147
Waterloo            60526
West Des Moines     60497
                    ...  
Rudd                   66
Albert City            62
Aurelia                55
Earling                42
Melbourne              36
Name: city, Length: 431, dtype: int64

The highest number of shops are located in Des Moines

Dealing with NaN`s and duplicates

In [12]:
test_task_data.dropna(inplace=True)
test_task_data.drop_duplicates()

invoice_and_item_number  ... volume_sold_gallons
0               INV-30655500013  ...                0.01
1               INV-30661000005  ...                0.19
2               INV-30659500001  ...                2.37
3               INV-30652600028  ...                0.19
4               INV-30650100014  ...                0.02
...                         ...  ...                 ...
1931584         INV-24313700016  ...                0.59
1931585         INV-24291000023  ...                3.17
1931586         INV-24313800006  ...                2.37
1931587         INV-24307700025  ...                3.17
1931588         INV-24303100045  ...                6.34

[1760082 rows x 24 columns]

Displaying of minimum and maximum dates in data

In [13]:
print(test_task_data['date'].min())
print(test_task_data['date'].max())

2020-01-02
2020-09-30


Displaying number and list of unique liquors in explored data

In [14]:
print(test_task_data['category_name'].nunique())
test_task_data['category_name'].unique()

57


array(['Blended Whiskies', 'American Brandies', 'Whiskey Liqueur',
       'American Dry Gins', 'Canadian Whiskies',
       'Straight Bourbon Whiskies', 'American Vodkas',
       'Tennessee Whiskies', 'Imported Cordials & Liqueurs',
       'Temporary & Specialty Packages', 'American Schnapps',
       'Imported Schnapps', 'American Cordials & Liqueur', 'White Rum',
       'Spiced Rum', 'Flavored Rum', 'Mixto Tequila',
       '100% Agave Tequila', 'Imported Flavored Vodka', 'Irish Whiskies',
       'Imported Vodkas', 'Scotch Whiskies', 'Cream Liqueurs',
       'Neutral Grain Spirits Flavored',
       'Temporary &  Specialty Packages', 'American Flavored Vodka',
       'Imported Brandies', 'Single Barrel Bourbon Whiskies',
       'Single Malt Scotch', 'Cocktails /RTD', 'Imported Dry Gins',
       'Imported Cordials & Liqueur', 'Coffee Liqueurs',
       'American Distilled Spirit Specialty', 'Bottled in Bond Bourbon',
       'Special Order Items', 'Aged Dark Rum', 'Triple Sec', 'Mezcal',
  

There are 57 unique liqors types in explored data

## Exploring the metrics which are important for business

**Top 10 categories of liquors that has highest sale**

In [15]:
df_plot = test_task_data.groupby(['category_name','pack','date'])['sale_dollars'].sum().groupby(
            ['category_name','pack']).max().sort_values().groupby(
            ['category_name']).sum().sort_values(ascending=False)
top_count = pd.DataFrame(df_plot)
top_count1 = pd.DataFrame(df_plot.head(10))

In [16]:
import plotly.graph_objects as go
fig_reg = px.bar(top_count1,x=top_count1.index, y='sale_dollars',color='sale_dollars')
fig_reg.update_layout(
    title="Sales of liquor per category",
    xaxis_title="Category Name",
    yaxis_title="Sales, dollars",
    )
fig_reg.show()

The highest number of sales in 2020 has American Vodkas

**Top 20 sales of liquor by city name**

In [51]:
df_plot = test_task_data.groupby(['city','pack','date'])['sale_dollars'].sum().groupby(
            ['city','pack']).max().sort_values().groupby(
            ['city']).sum().sort_values(ascending=False)
top_count1 = pd.DataFrame(df_plot)
top_count1 = pd.DataFrame(df_plot.head(20))

fig_reg = px.bar(top_count1,x=top_count1.index, y='sale_dollars',color='sale_dollars')
fig_reg.update_layout(
    title="Sales of liquor per city",
    xaxis_title="City Name",
    yaxis_title="Sales, dollars",
    )
fig_reg.show()

Des Moines city has the highest sales of liquors

**Sales of liquors per month**

In [17]:
test_task_data['date'] = pd.to_datetime(test_task_data['date'])
test_task_data['month'] = test_task_data['date'].dt.month
test_task_data['day'] = test_task_data['date'].dt.day
df_plot = test_task_data.groupby(['month','pack','date'])['sale_dollars'].sum().groupby(
            ['month','pack']).max().sort_values().groupby(
            ['month']).sum().sort_values(ascending=False)
top_count1 = pd.DataFrame(df_plot)
top_count1 = pd.DataFrame(df_plot.head(50))

fig_reg = px.bar(top_count1,x=top_count1.index, y='sale_dollars',color='sale_dollars')
fig_reg.update_layout(
    title="Sales of liquor per Month",
    xaxis_title=" Month Number",
    yaxis_title="Sales, dollars",
    )
fig_reg.show()

From the above plot it is cleared that the September month has highest number of sales of liquors

**Daily sales**

In [18]:
daily_sales = test_task_data.groupby('date', as_index=False)['sale_dollars'].sum()
daily_sales_sc = go.Scatter(x=daily_sales['date'], y=daily_sales['sale_dollars'])
layout = go.Layout(title='Daily sales', xaxis=dict(title='Date'), yaxis=dict(title='Sales'))
fig = go.Figure(data=[daily_sales_sc], layout=layout)
iplot(fig)

The highest daily sales are observed on 29 Sep 2020(First Trump-Biden debates)

**Sales of liquors with category and the volume of liquor sold**

In [19]:
df_plot = test_task_data.groupby(['category_name','pack','date'])['sale_dollars'].sum().groupby(
            ['category_name','pack']).max().sort_values().groupby(
            ['category_name']).sum().sort_values(ascending=False)
top_count1 = pd.DataFrame(df_plot)

df_plot = test_task_data.groupby(['category_name','pack','date'])['volume_sold_liters'].sum().groupby(
            ['category_name','pack']).max().sort_values().groupby(
            ['category_name']).sum().sort_values(ascending=False)
top_count2 = pd.DataFrame(df_plot)

top_count = pd.concat([top_count1 , top_count2],axis=1)
top_count = top_count.sort_values(['sale_dollars'],ascending=False)[:10]
top_count

sale_dollars  volume_sold_liters
American Vodkas                    665640.27            52512.73
Canadian Whiskies                  456244.56            37577.52
Imported Brandies                  403643.71             8269.75
Straight Bourbon Whiskies          326422.01            10393.24
Spiced Rum                         249269.31            16277.23
100% Agave Tequila                 199388.13             8533.27
Temporary & Specialty Packages     190692.82             7176.35
Whiskey Liqueur                    188888.20             9525.15
Imported Vodkas                    158407.96            10285.20
Tennessee Whiskies                 152309.34             5247.17

In [20]:
fig = go.Figure(data=[
    go.Bar(name='sale in dollars',x=top_count.index, y=top_count['sale_dollars']),
    go.Bar(name='Volume in litres',x=top_count.index, y=top_count['volume_sold_liters'])
])

fig.update_layout(barmode='group',title="Sales of liquors with category and the volume of liquor sold",
    xaxis_title="category",
    yaxis_title="Sale and the amount of liquor sold in litre",)
fig.show()

American Vodkas gas the highest sales $665640.27	with 52512.73 litre

**Volume sold in litres by vendor name**

In [21]:
df_plot = test_task_data.groupby(['vendor_name','pack','date'])['volume_sold_liters'].sum().groupby(
            ['vendor_name','pack']).max().sort_values().groupby(
            ['vendor_name']).sum().sort_values(ascending=False)
top_count1 = pd.DataFrame(df_plot)

fig_reg = px.bar(top_count1,x=top_count1.index, y='volume_sold_liters',color='volume_sold_liters')
fig_reg.update_layout(
    title="Volume sold by vendor name",
    xaxis_title=" Vendor Name",
    yaxis_title="Liquor sold in Litres",
    )
fig_reg.show()